File: 0053_maximumSubarray/main.cpp
#include <vector>
#include <algorithm>

class Solution {
public:
    int maxSubArray(std::vector<int>& nums) {
        if(nums.size() == 0)
            return 0;
        // dp[i] is the lenth of subarray that must contain ith number
        std::vector<int> dp(nums.size());
        dp[0] = nums[0];
        int max = dp[0];

        for(int i = 1; i < nums.size(); ++i) {
            if(dp[i-1] > 0) {
                dp[i] = dp[i-1] + nums[i];
            } else
                dp[i] = nums[i];

            if(dp[i] > max)
                max = dp[i];
        }
        return max;
    }
};

File: 0053_maximumSubarray/main.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 0:
            return 0

        dp = [None] * len(nums)
        dp[0] = nums[0]
        res = dp[0]

        for idx, num in enumerate(nums[1:], 1):
            dp[idx] = nums[idx] + (dp[idx-1] if dp[idx-1] > 0 else 0)
            res = max(dp[idx], res)
        
        return res

if __name__ == "__main__":
    test = [-2,1,-3,4,-1,2,1,-5,4]
    sol = Solution()
    sol.maxSubArray(test)